### PR TITLE
Rebase and merge of docker sysctl support

### PR DESCRIPTION
**Pull request by preetapan**

### CODE REVIEW

**Comment by commarla** [comment]
Thanks a lot @preetapan

File: website/source/docs/drivers/docker.html.md [hunk 1]
```diff
@@ -75,6 +75,7 @@ The `docker` driver supports the following configuration in the job spec.  Only
     }
     ```
 
+<<<<<<< 657619c0ec51068e48a34817ddc4df267b6cce6f
```
Comment (jippi) [inline website/source/docs/drivers/docker.html.md]: left-over merge conflict marker

Comment (preetapan) [reply website/source/docs/drivers/docker.html.md]: Fixed now

File: client/driver/docker.go [hunk 2]
```diff
@@ -199,6 +203,35 @@ type DockerDriverConfig struct {
...
+	for name, ulimitRaw := range ulimitsRaw {
+		// hard limit is optional
+		if strings.Contains(ulimitRaw, ":") == false {
+			ulimitRaw = ulimitRaw + ":" + ulimitRaw
+		}
+
+		splitted := strings.SplitN(ulimitRaw, ":", 2)
+		soft, err := strconv.Atoi(splitted[0])
+		if err != nil {
+			return []docker.ULimit{}, fmt.Errorf("Malformed ulimit %v: %v", name, ulimitRaw)
```
Comment (jippi) [inline client/driver/docker.go]: maybe include that it's the soft limit that is invalid? - same below for hard limit

File: client/driver/docker.go [hunk 3]
```diff
@@ -199,6 +203,35 @@ type DockerDriverConfig struct {
...
+	var ulimits []docker.ULimit
+
+	for name, ulimitRaw := range ulimitsRaw {
+		// hard limit is optional
+		if strings.Contains(ulimitRaw, ":") == false {
+			ulimitRaw = ulimitRaw + ":" + ulimitRaw
+		}
+
+		splitted := strings.SplitN(ulimitRaw, ":", 2)
+		soft, err := strconv.Atoi(splitted[0])
```
Comment (chelseakomlo) [inline client/driver/docker.go]: Check length before indexing.

File: client/driver/docker.go (hunk 2 above)
Comment (chelseakomlo) [reply client/driver/docker.go]: agreed with the limit specification for logging.

File: client/driver/docker.go [hunk 4]
```diff
@@ -199,6 +203,35 @@ type DockerDriverConfig struct {
...
+		if err != nil {
+			return []docker.ULimit{}, fmt.Errorf("Malformed ulimit %v: %v", name, ulimitRaw)
+		}
+		hard, err := strconv.Atoi(splitted[1])
```
Comment (chelseakomlo) [inline client/driver/docker.go]: Check limit before indexing.

File: client/driver/docker_test.go [hunk 5]
```diff
@@ -858,6 +858,60 @@ func TestDockerDriver_NetworkAliases_Bridge(t *testing.T) {
...
+	}
+
+	client, handle, cleanup := dockerSetup(t, task)
+	defer cleanup()
+
+	waitForExist(t, client, handle)
+
+	container, err := client.InspectContainer(handle.ContainerID())
+	if err != nil {
+		t.Fatalf("err: %v", err)
```
Comment (chelseakomlo) [inline client/driver/docker_test.go]: Use assert library for test assertions.

File: website/source/docs/drivers/docker.html.md [hunk 6]
```diff
@@ -97,6 +97,34 @@ The `docker` driver supports the following configuration in the job spec.  Only
 * `interactive` - (Optional) `true` or `false` (default). Keep STDIN open on
   the container.
 
+* `sysctl` - (Optional) A key-value map of sysctl configurations to set to the
```
Comment (chelseakomlo) [inline website/source/docs/drivers/docker.html.md]: This sentence is incomplete.

File: client/driver/docker_test.go [hunk 7]
```diff
@@ -858,6 +858,60 @@ func TestDockerDriver_NetworkAliases_Bridge(t *testing.T) {
 	}
 }
 
+func TestDockerDriver_Sysctl_Ulimit(t *testing.T) {
```
Comment (chelseakomlo) [inline client/driver/docker_test.go]: Should we test the failure conditions as well?

File: client/driver/docker_test.go [hunk 8]
```diff
@@ -858,6 +858,92 @@ func TestDockerDriver_NetworkAliases_Bridge(t *testing.T) {
...
+		task, _, _ := dockerTask(t)
+		task.Config["ulimit"] = tc.ulimitConfig
+
+		ctx := testDockerDriverContexts(t, task)
+		driver := NewDockerDriver(ctx.DriverCtx)
+		copyImage(t, ctx.ExecCtx.TaskDir, "busybox.tar")
+		defer ctx.AllocDir.Destroy()
+
+		if _, err := driver.Prestart(ctx.ExecCtx, task); err == nil || err.Error() != tc.err.Error() {
+			t.Fatalf("error expected in prestart, got %v, expected %v", err, tc.err)
```
Comment (chelseakomlo) [inline client/driver/docker_test.go]: Use assert here :)

**Comment by Ashald** [comment]
Is there any plans for supporting this for `rkt`?

**Comment by dadgar** [comment]
@Ashald Looks like there isn't support yet: https://github.com/rkt/rkt/issues/2694

**Comment by Ashald** [comment]
Sorry, missed that!

**Pull request merged**